### ISSUE TITLE
Make `implements/extends Function` a no-op.

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#33748]
This issue is concerned with taking the decision to make `implements Function` and `extends Function` have no effect, and subsequently updating the language specification to say so.

With Dart 2, an instance method named `call` does not any longer make a user-defined class a subtype of the function type which is the signature of that method.

Instead, in the situation where an instance of a type _T_ which is a class interface containing a member named `call` is used where a function type is expected, a tear-off of `call` is generated implicitly:

```dart
class C {
  void call(int i) {}
}

// OK, gives rise to implicit tear-off as in `... = new C().call;`.
void Function(int) f = new C();

// Compile-time error: Gives rise to implicit tear-off, but
// `void Function(int)` is not assignable to `void Function(int, int)`.
void Function(int, int) g = new C();
```

It was not uncommon in Dart 1 to declare a class with a `call` method using `implements Function`. That was redundant because such a class was made a subtype of the function type corresponding to the signature of the `call` method which makes it a subtype of `Function` per transitivity. In other words, `implements Function` was never useful, except perhaps as documentation.

In Dart 2, however, `implements Function` is directly misleading: When `o is Function` would yield `true` because `o` is an instance of a developer-written class with `implements Function`, `o` would in fact not be a function object, and there is no function type `F` such that `o` can be the value of a variable of type `F`. In that sense it is misleading to let a variable of type `Function` have `o` as its value: The tear-off of `call` should have taken place at some earlier point, that tear-off would have yielded an actual function object, and it would have the type `Function` as well as some function type. In that sense, `implements Function` just allows us to mask some bugs, consisting of a missing tear-off in the steps taken so far.

Still, it is a breaking change to go directly to a rule where `implements Function` is a compile-time error. So we may instead change the meaning of `implements Function` to have no effect!

Similarly, we should make `extends Function` have no effect, based on the same reasoning.

Granted, this is a very unusual choice, but it works for this particular situation because the implicit tear-off of a `call` method has no need for `implements Function`, it just requires a method named `call`.

**Comment by eernstg** [comment]
@leafpetersen, @lrhn, it is my impression that you support this decision. Please speak up if that is not true.

**Comment by lrhn** [comment]
I believe this was what we decided to do.

**Comment by eernstg** [comment]
This update handled via [this CL](https://dart-review.googlesource.com/c/sdk/+/88563).

**Comment by eernstg** [comment]
[Said CL](https://dart-review.googlesource.com/c/sdk/+/88563) landed as 68f77faa96200f275a9cf810f990f97c63ca583d.
